src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumeManager.java

```diff
@@ -0,0 +1,126 @@
+   *
+   * @param blobData the blob to upload
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage getStorage(List<List<ChannelData<T>>> blobData) {
+    // Only one chunk per blob in Iceberg mode.
+    ChannelFlushContext channelContext = blobData.get(0).get(0).getChannelContext();
+    StreamingIngestStorage stage =
+        this.externalVolumeMap.get(channelContext.getFullyQualifiedTableName());
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
+          String.format(
+              "No storage found for table %s", channelContext.getFullyQualifiedTableName()));
+    }
+
+    return stage;
+  }
+
+  /**
+   * Add a storage to the manager by looking up the table name from the open channel response
+   *
+   * @param openChannelResponse response from open channel
+   */
+  @Override
+  public void addStorage(OpenChannelResponse openChannelResponse) {
+    String fullyQualifiedTableName =
+        String.format(
+            "%s.%s.%s",
+            openChannelResponse.getDBName(),
+            openChannelResponse.getSchemaName(),
+            openChannelResponse.getTableName());
+    if (!this.externalVolumeMap.containsKey(fullyQualifiedTableName)) {
+      try {
+        ConfigureCallHandler configureCallHandler =
+            ConfigureCallHandler.builder(
+                    this.owningClient.getHttpClient(),
+                    this.owningClient.getRequestBuilder(),
+                    STREAMING_CHANNEL_CONFIGURE,
+                    CHANNEL_CONFIGURE_ENDPOINT)
+                .setRole(this.owningClient.getRole())
+                .setDatabase(openChannelResponse.getDBName())
+                .setSchema(openChannelResponse.getSchemaName())
+                .setTable(openChannelResponse.getTableName())
+                .build();
+        this.externalVolumeMap.put(
+            fullyQualifiedTableName,
+            new StreamingIngestStorage(
+                isTestMode,
+                configureCallHandler,
+                this.owningClient.getName(),
+                DEFAULT_MAX_UPLOAD_RETRIES));
+      } catch (SnowflakeSQLException | IOException err) {
+        throw new SFException(err, ErrorCode.UNABLE_TO_CONNECT_TO_STAGE);
+      }
+    }
+  }
+
+  // TODO: SNOW-1502887 Blob path generation for iceberg table
+  @Override
+  public String generateBlobPath() {
+    return "snow_dummy_file_name";
+  }
+
+  // TODO: SNOW-1502887 Blob path generation for iceberg table
+  @Override
+  public String getBlobPath(Calendar calendar, String clientPrefix) {
+    return "";
+  }
+
+  /**
+   * Get the client prefix from first external volume in the map
+   *
+   * @return the client prefix
+   */
+  @Override
+  public String getClientPrefix() {
+    if (this.externalVolumeMap.isEmpty()) {
+      return null;
+    }
+    return this.externalVolumeMap.values().iterator().next().getClientPrefix();
```

Comment:
lets chat about this, am not clear on what's being done here. Couple issues I see here:
1. return type should be Optional if there's a chance we'll return null
2. why is the first external volume "special"?
3. map.values doesn't typically guarantee what order it iterates in (for some map implementations if not all), so there's no guarantee that every call to map.values.iterator.next will give you the same object.
4. It appears that clientPrefix is a static value that isn't meant to change, so we should pass it in via the ctor if possible?

Reply:
Moved the `clientPrefix` logic to first configure call.